### PR TITLE
feature_1753:support kube-image-mirror-country and kube-image-registry

### PR DESCRIPTION
**Pull request by wuyingjun-lucky**
Signed-off-by: wuyingjun <wuyingjun_yewu@cmss.chinamobile.com>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
support kube-image-mirror-country and kube-image-registry
**Which issue(s) this PR fixes**:
Fixes #
https://github.com/karmada-io/karmada/issues/1753
**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by lonelyCZ** [comment]
> @lonelyCZ help to review

Ok, I will review it today.

File: pkg/karmadactl/cmdinit/cmdinit.go [hunk 1]
```diff
@@ -18,7 +18,7 @@ const (
 
 // NewCmdInit install karmada on kubernetes
 func NewCmdInit(cmdOut io.Writer, parentCommand string) *cobra.Command {
-	opts := kubernetes.CommandInitOption{}
+	opts := &kubernetes.CommandInitOption{}
```
Comment (lonelyCZ) [inline pkg/karmadactl/cmdinit/cmdinit.go:21]: Why using `&`?

File: pkg/karmadactl/cmdinit/kubernetes/deploy.go [hunk 2]
```diff
@@ -448,6 +468,49 @@ func (i *CommandInitOption) RunInit(_ io.Writer, parentCommand string) error {
...
+		value, ok := imageRepositories[mirrorCountry]
+		if ok {
+			return value
+		}
+	}
+
+	return imageRepositories["global"]
+}
+
+// get kube api server image
```
Comment (lonelyCZ) [inline pkg/karmadactl/cmdinit/kubernetes/deploy.go:490]:
```suggestion
// get kube-apiserver image
```

File: pkg/karmadactl/cmdinit/kubernetes/deploy.go [hunk 3]
```diff
@@ -71,6 +90,7 @@ type CommandInitOption struct {
 	CertAndKeyFileData                 map[string][]byte
 	RestConfig                         *rest.Config
 	KarmadaAPIServerIP                 []net.IP
+	FlagSet                            *pflag.FlagSet
```
Comment (lonelyCZ) [inline pkg/karmadactl/cmdinit/kubernetes/deploy.go:93]: FlagSet is not recommended as an option.

Comment (wuyingjun-lucky) [reply pkg/karmadactl/cmdinit/kubernetes/deploy.go]: @lonelyCZ  help to review again
I want to know if the user had set the kubeapiserver image manally.
three ways I have thought :
1: paas flagset to option and tell deploy.go if user had set the kubeapiserver image manally ( not recommended)
2: https://github.com/karmada-io/karmada/pull/1764/files#:~:text=cert%20and%20crds%20files%22),%22%22%2C%20%22Kubernetes%20apiserver%20image%22) set apiserver default value == "" and we use default value to judge user had set the value (code push now )
3: add  some options  in opts like apiserverset .  we set the apiserverset value by visit when flags init

**Comment by lonelyCZ** [comment]
> I want to know if the user had set the kubeapiserver image manally.

Yes, some users have their own image registry. 

> 2: https://github.com/karmada-io/karmada/pull/1764/files#:~:text=flags.StringVarP(%26opts.KarmadaAPIServerImage%2C%20%22karmada%2Dapiserver%2Dimage%22%2C%20%22%22%2C%20%22k8s.gcr.io/kube%2Dapiserver%3Av1.21.7%22%2C%20%22Kubernetes%20apiserver%20image%22)) set apiserver image default value == "" and we use default value to judge user had set the value (code push now )

This is good and easy. We no longer need this default value.

**Comment by wuyingjun-lucky** [comment]
> Yes, some users have their own image registry.
> 
> >
Ok， please review again @lonelyCZ

File: pkg/karmadactl/cmdinit/cmdinit.go [hunk 4]
```diff
@@ -108,6 +110,12 @@ func initExample(parentCommand string) string {
 # Private registry can be specified for all images` + "\n" +
 		fmt.Sprintf("%s init --etcd-image local.registry.com/library/etcd:3.5.1-0", parentCommand) + `
 
+# Kube registry can be specified by use kube-image-registry` + "\n" +
```
Comment (lonelyCZ) [inline pkg/karmadactl/cmdinit/cmdinit.go:113]:
```suggestion
# Kube registry can be specified by using kube-image-registry` + "\n" +
```

File: pkg/karmadactl/cmdinit/cmdinit.go [hunk 5]
```diff
@@ -108,6 +110,12 @@ func initExample(parentCommand string) string {
...
+		fmt.Sprintf("%s init --kube-image-registry=registry.cn-hangzhou.aliyuncs.com/google_containers", parentCommand) + `
+
+# China mainland registry mirror can be specified by use kube-image-mirror-country` + "\n" +
```
Comment (lonelyCZ) [inline pkg/karmadactl/cmdinit/cmdinit.go:116]:
```suggestion
# China mainland registry mirror can be specified by using kube-image-mirror-country` + "\n" +
```

File: pkg/karmadactl/cmdinit/kubernetes/deploy.go [hunk 6]
```diff
@@ -448,6 +461,49 @@ func (i *CommandInitOption) RunInit(_ io.Writer, parentCommand string) error {
...
+// get kube components registry
+func (i *CommandInitOption) kubeRegistry() string {
+	registry := i.KubeImageRegistry
+	mirrorCountry := strings.ToLower(i.KubeImageMirrorCountry)
+
+	if registry != "" {
+		return registry
+	}
+
+	if registry == "" && mirrorCountry != "" {
```
Comment (lonelyCZ) [inline pkg/karmadactl/cmdinit/kubernetes/deploy.go:473]:
```suggestion
	if mirrorCountry != "" {
```

At this point, `registry` has been empty

**Comment by wuyingjun-lucky** [comment]
@lonelyCZ 
How can I squash the commits in gitlab 
or I must squash in my computer ？

**Comment by lonelyCZ** [comment]
> @lonelyCZ How can I squash the commits in gitlab or I must squash in my computer ？

I think so latter.

**Comment by wuyingjun-lucky** [comment]
> > @lonelyCZ How can I squash the commits in gitlab or I must squash in my computer ？
> 
> I think so latter.

image
The git commit message format is ok ？

**Comment by lonelyCZ** [comment]
> The git commit message format is ok ？

Thanks, it is ok to keep track of your contributions.

**Comment by lonelyCZ** [comment]
I just tested it that worked fine. Just a little advice, rank the two new use cases to the front that are second and third.

```
Examples:

# Install Karmada in Kubernetes cluster
# The karmada-apiserver binds the master node's IP by default
karmadactl init

# Specify the URL to download CRD tarball
karmadactl init --crds https://github.com/karmada-io/karmada/releases/download/v1.1.0/crds.tar.gz

# Specify the local CRD tarball
karmadactl init --crds /root/crds.tar.gz

# Use PVC to persistent storage etcd data
karmadactl init --etcd-storage-mode PVC --storage-classes-name {StorageClassesName}

# Use hostPath to persistent storage etcd data. For data security, only 1 etcd pod can run in hostPath mode
karmadactl init --etcd-storage-mode hostPath  --etcd-replicas 1

# Use hostPath to persistent storage etcd data but select nodes by labels
karmadactl init --etcd-storage-mode hostPath --etcd-node-selector-labels karmada.io/etcd=true

# Private registry can be specified for all images
karmadactl init --etcd-image local.registry.com/library/etcd:3.5.1-0

# Kube registry can be specified by using kube-image-registry
karmadactl init --kube-image-registry=registry.cn-hangzhou.aliyuncs.com/google_containers

# China mainland registry mirror can be specified by using kube-image-mirror-country
karmadactl init --kube-image-mirror-country=cn
```

**Comment by wuyingjun-lucky** [comment]
> I just tested it that worked fine. Just a little advice, rank the two new use cases to the front that are second and third.
> 
> ```
> Examples:
> 
> # Install Karmada in Kubernetes cluster
> # The karmada-apiserver binds the master node's IP by default
> karmadactl init
> 
> # Specify the URL to download CRD tarball
> karmadactl init --crds https://github.com/karmada-io/karmada/releases/download/v1.1.0/crds.tar.gz
> 
> # Specify the local CRD tarball
> karmadactl init --crds /root/crds.tar.gz
> 
> # Use PVC to persistent storage etcd data
> karmadactl init --etcd-storage-mode PVC --storage-classes-name {StorageClassesName}
> 
> # Use hostPath to persistent storage etcd data. For data security, only 1 etcd pod can run in hostPath mode
> karmadactl init --etcd-storage-mode hostPath  --etcd-replicas 1
> 
> # Use hostPath to persistent storage etcd data but select nodes by labels
> karmadactl init --etcd-storage-mode hostPath --etcd-node-selector-labels karmada.io/etcd=true
> 
> # Private registry can be specified for all images
> karmadactl init --etcd-image local.registry.com/library/etcd:3.5.1-0
> 
> # Kube registry can be specified by using kube-image-registry
> karmadactl init --kube-image-registry=registry.cn-hangzhou.aliyuncs.com/google_containers
> 
> # China mainland registry mirror can be specified by using kube-image-mirror-country
> karmadactl init --kube-image-mirror-country=cn
> ```
fixed it
image

**Comment by lonelyCZ** [comment]
Thanks, @wuyingjun-lucky 

/lgtm

**Review by RainbowMango — APPROVED** [review]
/approve

**Pull request merged**